### PR TITLE
[BEAM-23] DirectRunner: rewrite single-output ParDo to multi-output

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This is in service of simplifying the override and expansion for splittable and stateful `ParDo`.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @tgroh

File: runners/direct-java/pom.xml [hunk 1]
```diff
@@ -74,6 +74,9 @@
                 <dependency>org.apache.beam:beam-sdks-java-core</dependency>
                 <dependency>org.apache.beam:beam-runners-core-java</dependency>
               </dependenciesToScan>
+              <excludes>
+                <exclude>org.apache.beam.sdk.transforms.OldDoFnTest</exclude>
```
Comment (tgroh) [inline runners/direct-java/pom.xml]: ???

Comment (tgroh) [inline runners/direct-java/pom.xml]: At least add a comment explaining why and linking to a JIRA

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    DoFn<InputT, OutputT> fn = transform.getNewFn();
+    DoFnSignature signature = DoFnSignatures.getSignature(fn.getClass());
+    if (!signature.processElement().isSplittable()) {
+      return transform;
+    } else {
+      return new SplittableParDo(fn);
+    }
+  }
+
+  static class GbkThenStatefulParDo<K, InputT, OutputT>
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: Is this used?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformEvaluatorRegistry.java [hunk 3]
```diff
@@ -51,9 +51,6 @@ public static TransformEvaluatorRegistry defaultRegistry(EvaluationContext ctxt)
             .put(Read.Bounded.class, new BoundedReadEvaluatorFactory(ctxt))
             .put(Read.Unbounded.class, new UnboundedReadEvaluatorFactory(ctxt))
             .put(
-                ParDo.Bound.class,
-                new ParDoEvaluatorFactory<>(ctxt, new ParDoSingleEvaluatorHooks<>()))
-            .put(
                 ParDo.BoundMulti.class,
                 new ParDoEvaluatorFactory<>(ctxt, new ParDoMultiEvaluatorHooks<>()))
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformEvaluatorRegistry.java]: Can you inline these hooks?

File: runners/direct-java/pom.xml (hunk 1 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: There are some overspecified tests of aggregators that we don't need to support. I'll see if I can isolate what is wrong with them.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java (hunk 2 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: Left over from `git cherry-pick` then pruning. Removed.

File: runners/direct-java/pom.xml (hunk 1 above)
Comment (kennknowles) [inline runners/direct-java/pom.xml]: Done. I think it all works now without exclusion. The inherited `aggregators` field continues to be a bit of a pain.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformEvaluatorRegistry.java (hunk 3 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformEvaluatorRegistry.java]: Done

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoOverrideFactory.java [hunk 4]
```diff
@@ -17,19 +17,16 @@
...
+import org.apache.beam.sdk.values.TupleTagList;
 
 /**
- * A {@link PTransformOverrideFactory} that provides overrides for applications of a {@link ParDo}
- * in the direct runner. Currently overrides applications of <a
- * href="https://s.apache.org/splittable-do-fn">Splittable DoFn</a>.
+ * A {@link PTransformOverrideFactory} that overrides single-output {@link ParDo} to implement
+ * it in terms of multi-output {@link ParDo}.
  */
 class ParDoOverrideFactory<InputT, OutputT>
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoOverrideFactory.java]: This probably should be named something like `ParDoSingleToMultiFactory`

**Comment by kennknowles** [comment]
Made the name change. I'll get it green and merge.

**Pull request closed without merging**